src/sonar_widget/SonarPlot.cc

```diff
@@ -76,23 +85,24 @@ bool SonarPlot::isMotorStepChanged(const base::Angle& bearing) {
     base::Angle diffStep = bearing - lastSonar.bearings[0];
     diffStep.rad = fabs(diffStep.rad);
 
-    if (!motorStep.isApprox(diffStep)) {
+    if (!motorStep.isApprox(diffStep) && lastDiffStep.isApprox(diffStep) && !bearing.isApprox(beforeLastSonar.bearings[0])) {
         motorStep = diffStep;
         numSteps = round(M_PI * 2 / motorStep.rad);
+        lastDiffStep = diffStep;
         return true;
     }
 
+    lastDiffStep = diffStep;
     return false;
 }
 
 // add current beam to accumulated scanning sonar data
 void SonarPlot::addScanningData(const base::samples::Sonar& sonar) {
-    int id_beam = lastSonar.bearings[0].rad / motorStep.rad;
-    if (id_beam < 0)
-        id_beam = (numSteps - 1) + id_beam;
+    int id_beam = (numSteps / 2) + sonar.bearings[0].rad / motorStep.rad;
+    if (id_beam == numSteps) id_beam--;
```

Comment:
My guess is that the problem is on `numSteps / 2`. It would probably be better to do `round((M_PI + sonar.bearings[0].rad) / motorStep.rad)`.

`round()` here is definitely best to fix rounding errors (i.e. if `expectedNumSteps * bearings.rad == numSteps * bearings.rad + epsilon`)